src/iop/demosaic.c

```diff
@@ -985,6 +993,1201 @@ static void xtrans_markesteijn_interpolate(float *out, const float *const in,
+          // explicitly been assigned to max (which signifies a new
+          // red/blue pair)
+          if(max == 0.0f)
+          {
+            float (*const pix)[3] = &rgb[0][row - top][col - left];
+            const short *const hex = hexmap(row, col, allhex);
+            for(int c = 0; c < 6; c++)
+            {
+              const float val = pix[hex[c]][1];
+              if(min > val) min = val;
+              if(max < val) max = val;
+            }
+          }
+          gmin[row - top][col - left] = min;
+          gmax[row - top][col - left] = max;
+          // handle vertical red/blue pairs
+          switch((row - sgrow) % 3)
+          {
+            // hop down a row to second pixel in vertical pair
+            case 1:
+              if(row < mrow - 4) row++, col--;
+              break;
+            // then if not done with the row hop up and right to next
+            // vertical red/blue pair, resetting min/max
+            case 2:
+              min = FLT_MAX, max = 0.0f;
+              if((col += 2) < mcol - 4 && row > top + 3) row--;
+          }
+        }
+      }
+
+      /* Interpolate green horizontally, vertically, and along both diagonals: */
+      // need a 3 pixel border here as 3*hex[] can have a 3 unit offset
+      const int pad_g_interp = 3;
+      for(int row = top + pad_g_interp; row < mrow - pad_g_interp; row++)
+        for(int col = left + pad_g_interp; col < mcol - pad_g_interp; col++)
+        {
+          float color[8];
+          int f = FCxtrans(row, col, roi_in, xtrans);
+          if(f == 1) continue;
+          float (*const pix)[3] = &rgb[0][row - top][col - left];
+          const short *const hex = hexmap(row, col, allhex);
+          // TODO: these constants come from integer math constants in
+          // dcraw -- calculate them instead from interpolation math
+          color[0] = 0.6796875f * (pix[hex[1]][1] + pix[hex[0]][1])
+                     - 0.1796875f * (pix[2 * hex[1]][1] + pix[2 * hex[0]][1]);
+          color[1] = 0.87109375f * pix[hex[3]][1] + pix[hex[2]][1] * 0.13f
+                     + 0.359375f * (pix[0][f] - pix[-hex[2]][f]);
+          for(int c = 0; c < 2; c++)
+            color[2 + c] = 0.640625f * pix[hex[4 + c]][1] + 0.359375f * pix[-2 * hex[4 + c]][1]
+                           + 0.12890625f * (2 * pix[0][f] - pix[3 * hex[4 + c]][f] - pix[-3 * hex[4 + c]][f]);
+          for(int c = 0; c < 4; c++)
+            rgb[c ^ !((row - sgrow) % 3)][row - top][col - left][1]
+                = CLAMPS(color[c], gmin[row - top][col - left], gmax[row - top][col - left]);
+        }
+
+      /* Interpolate red and blue values for solitary green pixels:   */
+      const int pad_rb_g = 6;
+      for(int row = (top - sgrow + pad_rb_g + 2) / 3 * 3 + sgrow; row < mrow - pad_rb_g; row += 3)
+        for(int col = (left - sgcol + pad_rb_g + 2) / 3 * 3 + sgcol; col < mcol - pad_rb_g; col += 3)
+        {
+          float(*rfx)
```

Comment:
Uhm, this does look like a clip, does it not?

Reply:
This is just part of the Markesteijn approach, which I merely took over. This is part of your codebase for a little while already, so I doubt that it makes the difference here. Actually, if you look at your initial review comments, you'll see that the clipping in question was at the input stage, and there, I removed it with my commit from 2 September.

Reply:
github is showing the wrong diff, see screenshot.
![screenshot_20171029_212613](https://user-images.githubusercontent.com/88600/32146946-da9a176a-bcef-11e7-965c-b773dad3c3eb.png)

Reply:
Ah, thx for clarifying. 
But this is also not the clipping we discussed before, which was really at the input stage. (and that one, sorry, I removed it)
This clipping here is after chroma estimation. It's there to avoid values going out of range in areas of high color variations. As such, this is pretty much common practice in demosaicing and all other algorithms also have it in one form or the other. We could have a look how the demosaicing behaves removing it, but I remember that I put this one for a very specific reason.